src/librustc/infer/canonical/query_response.rs

```diff
@@ -308,11 +308,12 @@ impl<'cx, 'gcx, 'tcx> InferCtxt<'cx, 'gcx, 'tcx> {
         // ...also include the other query region constraints from the query.
         output_query_region_constraints.extend(
             query_response.value.region_constraints.iter().filter_map(|r_c| {
-                let &ty::OutlivesPredicate(k1, r2) = r_c.skip_binder(); // reconstructed below
-                let k1 = substitute_value(self.tcx, &result_subst, &k1);
-                let r2 = substitute_value(self.tcx, &result_subst, &r2);
-                if k1 != r2.into() {
-                    Some(ty::Binder::bind(ty::OutlivesPredicate(k1, r2)))
+                let ty::OutlivesPredicate(k1, r2) = r_c.skip_binder(); // reconstructed below
```

Comment:
All this binder shuffling feels a bit complex. I think I micro-optimized this, though not necessarily for *very* good reason. Mostly I think I was trying to avoid allocating a whole vector of substituted results and instead substitute them lazilly. I wonder if it'd be simpler to do:

```
let r_c = substitute_value(self.tcx, &result_substs, r_c);

// Screen out `'a: 'a` cases -- we skip the binder here but
// only care the inner values to one another, so they are still at
// consistent binding levels.
let ty::OutlivesPredicate(k1, r2) = r_c.skip_binder();
if k1 != r2.into() {
    Some(r_c)
} else {
    None
}
```